File: packages/react-native-nitro-filetoolkit/android/src/main/java/com/margelo/nitro/filetoolkit/HybridFileReader.kt
package com.margelo.nitro.filetoolkit

import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import java.io.File
import java.io.RandomAccessFile

internal class HybridFileReader(
  override val location: FileLocation,
  file: File,
) : HybridFileReaderSpec() {
  private val lock = Any()
  private val handle = RandomAccessFile(file, "r")
  private val byteCount = handle.length()
  private var currentPosition = 0L
  private var closed = false

  override val position: ULong
    get() = synchronized(lock) { currentPosition.toULong() }

  override fun read(maxByteCount: ULong): Promise<ReadChunk> {
    if (maxByteCount > Int.MAX_VALUE.toULong()) {
      throw FileToolkitException.resourceLimit("maxByteCount exceeds the platform limit")
    }
    return Promise.parallel {
      synchronized(lock) {
        requireOpen()
        val offset = currentPosition
        val requested = maxByteCount.toInt()
        val bytes = ByteArray(requested)
        val count = if (requested == 0) 0 else handle.read(bytes)
        val result = if (count <= 0) ByteArray(0) else bytes.copyOf(count)
        currentPosition += result.size
        ReadChunk(
          data = ArrayBuffer.copy(result),
          offset = offset,
          isEndOfFile = currentPosition >= byteCount,
        )
      }
    }
  }

  override fun seek(offset: ULong): Promise<Unit> = Promise.parallel {
    if (offset > Long.MAX_VALUE.toULong()) {
      throw FileToolkitException.resourceLimit("offset exceeds the platform limit")
    }
    synchronized(lock) {
      requireOpen()
      handle.seek(offset.toLong())
      currentPosition = offset.toLong()
    }
  }

  override fun close() {
    synchronized(lock) {
      if (closed) return
      handle.close()
      closed = true
    }
  }

  private fun requireOpen() {
    if (closed) throw FileToolkitException.invalidOperation("reader is closed")
  }
}

File: packages/react-native-nitro-filetoolkit/ios/HybridFileReader.swift
import Foundation
import NitroModules

final class HybridFileReader: HybridFileReaderSpec {
  let location: FileLocation

  private let queue = DispatchQueue(label: "com.margelo.nitro.filetoolkit.reader", qos: .utility)
  private let lock = NSLock()
  private let handle: FileHandle
  private let byteCount: UInt64
  private var currentPosition: UInt64 = 0
  private var isClosed = false

  var position: UInt64 {
    lock.lock()
    defer { lock.unlock() }
    return currentPosition
  }

  init(location: FileLocation, url: URL) throws {
    self.location = location
    handle = try FileHandle(forReadingFrom: url)
    let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
    byteCount = (attributes[.size] as? NSNumber)?.uint64Value ?? 0
    super.init()
  }

  func read(maxByteCount: UInt64) throws -> Promise<ReadChunk> {
    guard maxByteCount <= UInt64(Int.max) else {
      throw FileToolkitError.limitExceeded("maxByteCount exceeds the platform limit")
    }
    return Promise.parallel(queue) { [self] in
      lock.lock()
      defer { lock.unlock() }
      try requireOpen()
      let offset = currentPosition
      let data = try handle.read(upToCount: Int(maxByteCount)) ?? Data()
      currentPosition += UInt64(data.count)
      return ReadChunk(
        data: try ArrayBuffer.copy(data: data),
        offset: offset,
        isEndOfFile: currentPosition >= byteCount
      )
    }
  }

  func seek(offset: UInt64) throws -> Promise<Void> {
    Promise.parallel(queue) { [self] in
      lock.lock()
      defer { lock.unlock() }
      try requireOpen()
      try handle.seek(toOffset: offset)
      currentPosition = offset
    }
  }

  func close() throws {
    lock.lock()
    defer { lock.unlock() }
    guard !isClosed else { return }
    try handle.close()
    isClosed = true
  }

  private func requireOpen() throws {
    if isClosed {
      throw FileToolkitError.invalidOperation("reader is closed")
    }
  }

  deinit {
    try? handle.close()
  }
}
